onnxruntime/core/providers/cpu/ml/treeregressor.cc

```diff
@@ -239,40 +287,172 @@ common::Status TreeEnsembleRegressor<T>::Compute(OpKernelContext* context) const
   int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
   Tensor* Y = context->Output(0, TensorShape({N, n_targets_}));
 
-  int64_t write_index = 0;
   const auto* x_data = X->template Data<T>();
 
-  for (int64_t i = 0; i < N; i++)  //for each class
-  {
-    int64_t current_weight_0 = i * stride;
-    std::unordered_map<int64_t, std::tuple<float, float, float>> scores; // sum, min, max
-    //for each tree
-    for (size_t j = 0; j < roots_.size(); j++) {
-      //walk each tree from its root
-      ORT_RETURN_IF_ERROR(ProcessTreeNode(scores, roots_[j], x_data, current_weight_0));
+  if (n_targets_ == 1) {
+    float origin = base_values_.size() == 1 ? base_values_[0] : 0.f;
+    if (N == 1) {
+      float scores = 0;
+      unsigned char has_scores = 0;
+
+      if (nbtrees_ >= 2 && (aggregate_function_ == AGGREGATE_FUNCTION::AVERAGE ||
+                            aggregate_function_ == AGGREGATE_FUNCTION::SUM)) {
+#ifdef USE_OPENMP
+#pragma omp parallel for reduction(|                         \
+                                   : has_scores) reduction(+ \
+                                                           : scores)
+#endif
+        for (int64_t j = 0; j < nbtrees_; ++j)
+          ProcessTreeNode(&scores, roots_[j], x_data, &has_scores);
+      } else {
+        for (int64_t j = 0; j < nbtrees_; ++j)
+          ProcessTreeNode(&scores, roots_[j], x_data, &has_scores);
+      }
+
+      float val = has_scores
+                      ? (aggregate_function_ == AGGREGATE_FUNCTION::AVERAGE
+                             ? scores / roots_.size()
+                             : scores) +
+                            origin
+                      : origin;
+      *((float*)(Y->template MutableData<T>())) = (post_transform_ == POST_EVAL_TRANSFORM::PROBIT)
+                                                      ? ComputeProbit(val)
```

Comment:
Is there a reason to not use T for scores and val so this cast isn't necessary?

Reply:
This is tricky: ONNX specification mentions threshold are float but input and output of the node can be float or doubles. That means there are two types, float for threshold, float or double for targets. My implementation is templated, that means both types can be changed even though the first one does not. However, the best design would be all floats or all doubles.